Assets/Prefabs/Player/PlayerController.cs

```diff
@@ -0,0 +1,173 @@
+    public float walkSpeed = 5.0f;
+    [Tooltip("Amount of time it takes to reach full walk / run speed (sec)")]
+    public float speedUpTime = 0.2f;
+    [Tooltip("Time it takes for the character to stop when the joystick is released (sec)")]
+    public float stopTime = 0.1f;
+    // TODO: IMPLEMENT THIS IN TERMS OF TIME BY SCALING WITH TIME.DELTATIME
+    [Tooltip("Portion of the rotatin between curent and desired orientation to apply each physics update"), Range(0, 1.0f)]
+    public float rotationRatio = 0.2f;
+    [Tooltip("Gravitational Constant (units / sec / sec)")]
+    public float gravity = -10.0f;
+
+    private CharacterController characterController;
+    private float verticalVelocity = 0.0f;
+    private float horizontalSpeed = 0.0f;
+    private float horizontalAcceleration = 0.0f;
+
+    /// <summary>
+    /// Get a reference to the GameObject's <c>Charactercontroller</c> <c>Component</c>
+    /// for use
+    /// </summary>
+    /// <remarks>Must be called during <c>OnAwake()</c></remarks>
+    private void OnAwakeGetCharacterController()
+    {
+        characterController = GetComponent<CharacterController>();
+    }
+
+    /// <summary>
+    /// Apply motion to the GameObject based on player input, the current camera 
+    /// orientation, and gravity.
+    /// </summary>
+    /// <remarks>
+    /// <list type="bullet">
+    /// <item>
+    /// Motion is performed by first rotating the object towards the direction of player 
+    /// input (relative to the current camera view) <em>then</em> moving along that 
+    /// direction.
+    /// </item>
+    /// <item>
+    /// Rotation is achieved by using a spherical linear-interpolation (SLERP) between 
+    /// the current object orientation Quaternion and the "dired direction of travel"
+    /// Quaternion.
+    /// </item>
+    /// <item>
+    /// Forward motion is achieved using a smooth-damping function to approach the 
+    /// maximum walking speed scaled by the player's input.
+    /// </item>
+    /// </list>
+    /// </remarks>
+    private void ApplyMotion()
+    {
+        if (characterController != null)
+        {
+            if (moveMagnitude > 0)
+            {
+                Vector3 directionOfPlayerInput = Quaternion.AngleAxis(moveAngle, Vector3.up) * MyUtils.ProjectForwardOntoXZPlane(myCamera.transform);
+                Quaternion orientationOfPlayerInput = Quaternion.LookRotation(directionOfPlayerInput, Vector3.up);
+                transform.rotation = Quaternion.Slerp(transform.rotation, orientationOfPlayerInput, rotationRatio);
+
+                horizontalSpeed = Mathf.SmoothDamp(horizontalSpeed, walkSpeed * moveMagnitude, ref horizontalAcceleration, speedUpTime);
+            }
+            else
+            {
+                horizontalSpeed = Mathf.SmoothDamp(horizontalSpeed, 0.0f, ref horizontalAcceleration, stopTime);
+            }
+            characterController.Move(horizontalSpeed * Time.deltaTime * transform.forward);
```

Comment:
I see, character controller is moving horizontally as defined by the controller, it doesnt use the physics engine in any way ? The vertical position depends on physics due to gravity.

Reply:
Oooooo this is a fun one 😄 

So there are a lot of really good videos on the topic of Character Controllers ([like this one on on of my favoritest platformers ever](https://www.youtube.com/watch?v=yorTG9at90g)) - but this short version is that using realistic physics for the player character feels _gross_ and _sticky_. Things move slow and wonky, and it doesn't feelright.

So Unity has this nifty [`CharacterController` Component](https://docs.unity3d.com/Manual/CharacterControllers.html) that is specifically designed to let you use the Physics engine for things that matter (walls, floors, collision boxes, etc.) but have full control over the motion of your object without trying to emulate quick movements with acceleration profiles and all sorts of wacky stuff!